compiler/rustc_session/src/options.rs

```diff
@@ -835,6 +837,38 @@ mod parse {
         true
     }
 
+    crate fn parse_prefer_dynamic(slot: &mut PreferDynamicSet, v: Option<&str>) -> bool {
+        let s = match v {
+            // Note: n/no/off do *not* map to an empty-set of crates.
+            //
+            // This is to continue supporting rustc's historical behavior where it attempts to
+            // link everything statically, but failing that, then greedily link as many crates
+            // dynamically as it can.
+            //
+            // If these options mapped to an empty set, then that would denote that no dynamic
+            // linkage should be given preference over static, which would not correspond to
+            // historical meaning of `-C prefer-dynamic=no`.
+            //
+            // (One requests an empty set by writing `-C prefer-dynamic=`, with an empty string
+            // as the value.)
+            Some("n") | Some("no") | Some("off") => PreferDynamicSet::unset(),
+
+            // `-C prefer-dynamic` gives all crates preferred dynamic linkage.
+            // `-C prefer-dynamic=...` with `y`/`yes`/`on` is a synonym, for backwards
+            // compatibility.
+            Some("y") | Some("yes") | Some("on") | None => PreferDynamicSet::every_crate(),
+
+            // `-C prefer-dynamic=crate1,crate2,...` gives *just* crate1, crate2, ... preferred
+            // dynamic linkage.
+            Some(s) => {
+                let v = s.split(',').map(|s| s.to_string()).collect();
+                PreferDynamicSet::Set(v)
+            }
```

Comment:
Maybe it should be a different argument as the new option is not a preference, but mandatory?

Reply:
Hmm. My intention was that it would still be a preference, in the sense that if you do `-C prefer-dynamic=crate1`, but only a static library for `crate1` is available, then linkage will still proceed using the static library.

I probably should add a test illustrating that.

Reply:
On reflection, I do think I must have messed up something in my logic that ends up making *something* a requirement rather than a preference. Specifically, I'm working on a run-make test analogous to https://github.com/rust-lang/rust/blob/master/src/test/run-make-fulldeps/dylib-chain/Makefile
but where I choose variations of whether dylibs on the chain have both rlib and dylib, or just rlib, or just dylib.

The problem I'm seeing is that in some scenarios using `-C prefer-dynamic=crate,...`, a crate that is solely provided as a dylib is not getting linked in unless I include it in the above list. That wasn't my intention here: My intention was that the preference is meant to give guidance when the compiler must make a choice, not act as an assertion about what it expects to be present.

Reply:
Update 2: Hmm maybe I was wrong about the above: In general `.so` files don't have enough info to be statically linked. But then I'm really confused about some of the behavior I'm seeing. Still looking.

Reply:
Okay, I have now made a run-make test that I think illustrates the current behavior of the system, including cases where rlib's end up depending on dylibs.

I don't think pre-existing behavior is necessarily *intuitive*, but my main goal right now is to address missing use cases, **not** to try to figure out what the ideal behavior is for all cases and potentially inject breaking changes to support that idealized behavior.